vendor/sigs.k8s.io/cloud-provider-azure/pkg/batch/batch.go

```diff
@@ -0,0 +1,259 @@
+}
+
+// newEntry returns a new batch entry.
+func newEntry(ctx context.Context, value interface{}) *entry {
+	return &entry{ctx: ctx, value: value, resultChan: make(chan Result, 1)}
+}
+
+// setResult sets the result of a batch entry.
+func (e *entry) setResult(value interface{}, err error) {
+	e.resultChan <- Result{value: value, err: err}
+	close(e.resultChan)
+}
+
+// getResult waits for and returns the result of a batch entry.
+func (e *entry) getResult(ctx context.Context) (interface{}, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+
+	case r := <-e.resultChan:
+		return r.value, r.err
+	}
+}
+
+// getResultChar returns the result channel of a batch entry.
+func (e *entry) getResultChan() <-chan (Result) {
+	return e.resultChan
+}
+
+// add adds a value to the batch.
+func (b *batcher) add(ctx context.Context, value interface{}) (e *entry, first bool) {
+	e = newEntry(ctx, value)
+
+	b.mutex.Lock()
+	first = len(b.entries) == 0
+	b.entries = append(b.entries, e)
+	b.mutex.Unlock()
+
+	return
+}
+
+// flush returns the current batch of values and empties the batch. It also returns the time of the
+// furthest deadline in the batch. If no entries specify a deadline in their context, deadline.IsZero() == true.
+func (b *batcher) flush() (flushed []*entry, nextDeadline time.Time) {
+	var entries []*entry
+
+	b.mutex.Lock()
+	entries = b.entries
+	b.entries = make([]*entry, 0)
+	b.mutex.Unlock()
+
+	for _, entry := range entries {
+		if ctxErr := entry.ctx.Err(); ctxErr != nil {
+			entry.setResult(nil, ctxErr)
+		} else {
+			flushed = append(flushed, entry)
+
+			if deadline, ok := entry.ctx.Deadline(); ok && deadline.After(nextDeadline) {
+				nextDeadline = deadline
+			}
+		}
+	}
+
+	return
+}
+
+// purgeExpiredOrCanceled removes any expired or canceled entries from the batch. It returns true and
+// the furthest deadline in the remaining batch if the batch is not empty afterward (deadline.IsZero() == true
+// none specify a deadline). It returns false if the batch is empty after removal of expired or canceled
+// entries.
+func (b *batcher) purgeExpiredOrCanceled() (more bool, nextDeadline time.Time) {
+	kept := make([]*entry, 0)
+
+	b.mutex.Lock()
+	for _, entry := range b.entries {
+		if ctxErr := entry.ctx.Err(); ctxErr != nil {
+			entry.setResult(nil, ctxErr)
+		} else {
+			kept = append(kept, entry)
+
+			if deadline, ok := entry.ctx.Deadline(); ok && deadline.After(nextDeadline) {
+				nextDeadline = deadline
+			}
+		}
+	}
+
+	more = len(kept) > 0
+	b.entries = kept
+	b.mutex.Unlock()
+
+	return
+}
+
+// tryEnterProcessLoop returns true if the batch processing function can be invoked by this goroutine.
+// This function ensures that only one goroutine at any given time can enter the batch processing loop.
+//
+// The returned function must be called on exit from the processing loop.
+func (b *batcher) tryEnterProcessLoop() (bool, func()) {
+	deadline := time.Now().Add(2 * time.Minute)
+
+	for {
```

Comment:
Do we need this loop? It seems like we would always return on the first iteration.

Reply:
Oh, I see. I missed the later part.